test/cpp/end2end/tls_private_key_offload_end2end_test.cc

```diff
@@ -0,0 +1,422 @@
+  client_identity_key_cert_pairs.emplace_back(client_key_cert_pair);
+  auto client_certificate_provider =
+      std::make_shared<experimental::StaticDataCertificateProvider>(
+          ca_cert, client_identity_key_cert_pairs);
+  grpc::experimental::TlsChannelCredentialsOptions options;
+  options.set_certificate_provider(client_certificate_provider);
+  options.watch_root_certs();
+  options.set_root_cert_name("root");
+  options.watch_identity_key_cert_pairs();
+  options.set_identity_cert_name("identity");
+  options.set_check_call_host(false);
+
+  DoRpc(server_addr_, options);
+}
+
+TEST_F(TlsPrivateKeyOffloadTest, OffloadWithCustomKeySigner) {
+  server_addr_ = absl::StrCat("localhost:",
+                              std::to_string(grpc_pick_unused_port_or_die()));
+  std::string server_key =
+      grpc_core::testing::GetFileContents(std::string(kServerKeyPath));
+  std::string server_cert =
+      grpc_core::testing::GetFileContents(std::string(kServerCertPath));
+  std::string ca_cert =
+      grpc_core::testing::GetFileContents(std::string(kCaPemPath));
+
+  auto server_certificate_provider =
+      std::make_shared<experimental::InMemoryCertificateProvider>();
+  signer_ = std::make_shared<TestCustomPrivateKeySigner>(server_key);
+  grpc_core::PemKeyCertPairList identity_pairs;
+  identity_pairs.emplace_back(signer_, server_cert);
+  static_cast<grpc_core::InMemoryCertificateProvider*>(
+      server_certificate_provider->c_provider())
+      ->UpdateIdentity(identity_pairs);
+  static_cast<grpc_core::InMemoryCertificateProvider*>(
+      server_certificate_provider->c_provider())
+      ->UpdateRoot(std::make_shared<RootCertInfo>(ca_cert));
+
+  absl::Notification notification;
+  server_thread_ = new std::thread(
+      [&]() { RunServer(&notification, server_certificate_provider); });
+  notification.WaitForNotification();
+
+  std::string client_key =
+      grpc_core::testing::GetFileContents(std::string(kClientKeyPath));
+  std::string client_cert =
+      grpc_core::testing::GetFileContents(std::string(kClientCertPath));
+  experimental::IdentityKeyCertPair key_cert_pair;
+  key_cert_pair.private_key = client_key;
+  key_cert_pair.certificate_chain = client_cert;
+  std::vector<experimental::IdentityKeyCertPair> identity_key_cert_pairs;
+  identity_key_cert_pairs.emplace_back(key_cert_pair);
+  auto client_certificate_provider =
+      std::make_shared<experimental::StaticDataCertificateProvider>(
+          ca_cert, identity_key_cert_pairs);
+  grpc::experimental::TlsChannelCredentialsOptions options;
+  options.set_certificate_provider(client_certificate_provider);
+  options.watch_root_certs();
+  options.set_root_cert_name("root");
+  options.watch_identity_key_cert_pairs();
+  options.set_identity_cert_name("identity");
+  options.set_check_call_host(false);
+
+  DoRpc(server_addr_, options);
+}
+
+TEST_F(TlsPrivateKeyOffloadTest, OffloadWithCustomKeySignerCStyleApi) {
```

Comment:
This test should be removed.  This file is a C++ test, so it should test the C++ API, not the internal C-core API.

If we need more coverage of the C-core API, then that test should be somewhere in test/core.